<html>
 <head>
  <title>Report for Sample2</title>
 </head>
<body>
<p align="center"><h1>Report for Sample2</h1></p>
<p>Sex determined from SRY MIPs: <strong>F (SRY/Total: 0/8835 = 0.00%)</strong></p>
<p>Number of target region performance outlier MIPs: <strong>0</strong></p>
<p>Number of sites with GTs: <strong>12932 (100.00%)</strong></p>
<p>Number of sites with GTs [F8]: <strong>9846 (100.00%)</strong></p>
<p>Number of sites with GTs [F9]: <strong>3086 (100.00%)</strong></p>
<p>Average coverage: <strong>45.56</strong></p>
<p>Average coverage of GT calls: <strong>45.56</strong></p>
<p>Number of hets called: <strong>0</strong></p>
<p>Number of short variants called: <strong>1</strong></p>
<p>Number of low quality variants not counted above: <strong>0</strong></p>
<p></p>
<p><h2>Variants identified in target region:</h2></p>
<table cellpadding="5" border="3">
<tr><th>GT</th><th>GQ</th><th>AD</th><th>DP</th><th>Status</th><th>Chrom</th><th>Start</th><th>End</th><th>Allele</th><th>Gene</th><th>Feature</th><th>Feature<br>type</th><th>Consequence</th><th>cDNA<br>position</th><th>CDS<br>position</th><th>Protein<br>position</th><th>Amino<br>acids</th><th>Codons</th><th>Existing<br>variation</th><th>Extra</th></tr>
<tr><td >1/1</td><td >99</td><td >0,55</td><td >55</td><td >OK</td><td >X</td><td >154194909</td><td >154194909</td><td >A</td><td >ENSG00000185010</td><td >ENST00000360256</td><td >Transcript</td><td >stop_gained</td><td >1264</td><td >1063</td><td >355</td><td >R/*</td><td >Cga/Tga</td><td >rs137852368<br>CM880025</td><td >IMPACT=HIGH<br>STRAND=-1<br>SYMBOL=F8<br>SYMBOL_SOURCE=HGNC<br>HGNC_ID=3546<br>CANONICAL=YES<br>CCDS=CCDS35457.1<br>EXON=8/26<br>DOMAINS=PIRSF_domain:PIRSF000354<br>HGVSc=ENST00000360256.4:c.1063C>T<br>HGVSp=ENSP00000353393.4:p.Arg355Ter<br>CLIN_SIG=pathogenic<br>PHENO=1,1</td></tr>
</table>
<p>GT - Genotype call encoded as allele values separated by "/". The allele values are 0 for the reference allele, 1 for the first alternative allele, 2 for the second allele.<br>
GQ - Conditional genotype quality, encoded as a phred quality -10*log_10(p) (genotype call is wrong, conditioned on the site's being variant); we only report GQ >= 30 on the summary page.<br>
AD - Read depth for each variant at this position (first reference, followed by alternative alleles).<br>
DP - Read depth at this position for this sample; we only report DP >= 8 on the summary page.</p>
<p></p>
<p><h2>Coverage by target region</h2></p>
<table cellpadding="5" border="3">
<tr><th>Region</th><th>Length</th><th>Sites</th><th>Called</th><th>Fraction</th><th>Ave.Cov</th></tr>
<tr><td>F8/upstream</td><td>279</td><td>279</td><td>279</td><td>100.00%</td><td>38.30</td></tr>
<tr><td>F8/5-UTR</td><td>171</td><td>171</td><td>171</td><td>100.00%</td><td>34.64</td></tr>
<tr><td>F8/1</td><td>153</td><td>153</td><td>153</td><td>100.00%</td><td>33.09</td></tr>
<tr><td>F8/2</td><td>163</td><td>163</td><td>163</td><td>100.00%</td><td>91.56</td></tr>
<tr><td>F8/3</td><td>143</td><td>143</td><td>143</td><td>100.00%</td><td>73.87</td></tr>
<tr><td>F8/4</td><td>244</td><td>244</td><td>244</td><td>100.00%</td><td>41.04</td></tr>
<tr><td>F8/5</td><td>89</td><td>89</td><td>89</td><td>100.00%</td><td>36.15</td></tr>
<tr><td>F8/6</td><td>143</td><td>143</td><td>143</td><td>100.00%</td><td>99.00</td></tr>
<tr><td>F8/7</td><td>251</td><td>251</td><td>251</td><td>100.00%</td><td>31.33</td></tr>
<tr><td>F8/8</td><td>282</td><td>282</td><td>282</td><td>100.00%</td><td>56.82</td></tr>
<tr><td>F8/9</td><td>204</td><td>204</td><td>204</td><td>100.00%</td><td>65.92</td></tr>
<tr><td>F8/10</td><td>124</td><td>124</td><td>124</td><td>100.00%</td><td>28.94</td></tr>
<tr><td>F8/11</td><td>247</td><td>247</td><td>247</td><td>100.00%</td><td>72.90</td></tr>
<tr><td>F8/12</td><td>171</td><td>171</td><td>171</td><td>100.00%</td><td>65.64</td></tr>
<tr><td>F8/13</td><td>230</td><td>230</td><td>230</td><td>100.00%</td><td>69.33</td></tr>
<tr><td>F8/14</td><td>3126</td><td>3119</td><td>3119</td><td>100.00%</td><td>52.61</td></tr>
<tr><td>F8/15</td><td>181</td><td>181</td><td>181</td><td>100.00%</td><td>45.31</td></tr>
<tr><td>F8/16</td><td>233</td><td>233</td><td>233</td><td>100.00%</td><td>53.01</td></tr>
<tr><td>F8/17</td><td>249</td><td>249</td><td>249</td><td>100.00%</td><td>24.98</td></tr>
<tr><td>F8/18</td><td>203</td><td>203</td><td>203</td><td>100.00%</td><td>24.52</td></tr>
<tr><td>F8/19</td><td>137</td><td>137</td><td>137</td><td>100.00%</td><td>59.11</td></tr>
<tr><td>F8/20</td><td>92</td><td>92</td><td>92</td><td>100.00%</td><td>32.12</td></tr>
<tr><td>F8/21</td><td>106</td><td>106</td><td>106</td><td>100.00%</td><td>31.75</td></tr>
<tr><td>F8/22</td><td>176</td><td>176</td><td>176</td><td>100.00%</td><td>35.41</td></tr>
<tr><td>F8/23</td><td>174</td><td>174</td><td>174</td><td>100.00%</td><td>52.28</td></tr>
<tr><td>F8/24</td><td>169</td><td>169</td><td>169</td><td>100.00%</td><td>33.17</td></tr>
<tr><td>F8/25</td><td>197</td><td>197</td><td>197</td><td>100.00%</td><td>58.39</td></tr>
<tr><td>F8/26</td><td>166</td><td>166</td><td>166</td><td>100.00%</td><td>14.17</td></tr>
<tr><td>F8/3-UTR</td><td>1808</td><td>1720</td><td>1720</td><td>100.00%</td><td>26.92</td></tr>
<tr><td>F8/downstream</td><td>30</td><td>30</td><td>30</td><td>100.00%</td><td>13.63</td></tr>
<tr><td>F9/upstream</td><td>271</td><td>271</td><td>271</td><td>100.00%</td><td>57.13</td></tr>
<tr><td>F9/5-UTR</td><td>29</td><td>29</td><td>29</td><td>100.00%</td><td>71.24</td></tr>
<tr><td>F9/1</td><td>98</td><td>98</td><td>98</td><td>100.00%</td><td>79.00</td></tr>
<tr><td>F9/2</td><td>184</td><td>184</td><td>184</td><td>100.00%</td><td>60.15</td></tr>
<tr><td>F9/3</td><td>45</td><td>45</td><td>45</td><td>100.00%</td><td>62.42</td></tr>
<tr><td>F9/4</td><td>141</td><td>141</td><td>141</td><td>100.00%</td><td>57.48</td></tr>
<tr><td>F9/5</td><td>173</td><td>173</td><td>173</td><td>100.00%</td><td>32.72</td></tr>
<tr><td>F9/6</td><td>236</td><td>236</td><td>236</td><td>100.00%</td><td>48.39</td></tr>
<tr><td>F9/7</td><td>135</td><td>135</td><td>135</td><td>100.00%</td><td>35.45</td></tr>
<tr><td>F9/8</td><td>558</td><td>558</td><td>558</td><td>100.00%</td><td>40.89</td></tr>
<tr><td>F9/3-UTR</td><td>1387</td><td>1186</td><td>1186</td><td>100.00%</td><td>36.56</td></tr>
<tr><td>F9/downstream</td><td>30</td><td>30</td><td>30</td><td>100.00%</td><td>48.00</td></tr>
</table>
<p></p>
<p><h2>Regions with missing genotype calls</h2></p>
<p>None</p>
<p></p>
<p><h2>Inversion MIP results</h2></p>
<p>Total inversion MIP reads: <strong>0</strong></p>
<p><h3>Resulting inversion calls</h3></p>
<p><em>INT1-FAILED</em></p>
<p><em>INT22-FAILED</em></p>
<p></p>
<p><h2>Under-/over-performing target-region MIPs</h2></p>
<p>None</p>
<p>&nbsp;</p>
<p>Go to <a href="report.html">Sample Summary Report</a></p>
<p>Go to <a href="summary.html">Summary Report</a></p>
</body>
</html>
